app/src/main/java/com/example/taskyy/data/repositories/UserPreferencesImpl.kt

```diff
@@ -0,0 +1,57 @@
+package com.example.taskyy.data.repositories
+
+import android.content.SharedPreferences
+import com.example.taskyy.domain.objects.AuthenticatedUser
+import com.example.taskyy.domain.repository.UserPreferences
+import com.google.gson.Gson
+
+class UserPreferencesImpl(
+    private val sharedPref: SharedPreferences
+) : UserPreferences {
+    override fun saveAuthenticatedUser(user: AuthenticatedUser, key: String) {
+        val gson = Gson()
+        val json = gson.toJson(user)
+        sharedPref.edit().putString(key, json).apply()
+    }
+
+    override fun clearPreferences() {
+        sharedPref.edit().clear().apply()
+    }
+
+    override fun addUserFullName(fullName: String, key: String) {
+        sharedPref.edit().putString(fullName, key).apply()
+    }
+
+    override fun getUserFullName(key: String): String {
+        val name = sharedPref.getString(key, "")
+        return name!!
+    }
+
+    override fun addUserEmail(email: String, key: String) {
+        sharedPref.edit().putString(email, key).apply()
+
+    }
+
+    override fun getUserEmail(key: String): String {
+        val email = sharedPref.getString(key, "")
+        return email!!
+    }
+
+    override fun addUserId(userId: String, key: String) {
+        sharedPref.edit().putString(userId, key).apply()
+    }
+
+    override fun getUserId(key: String): String {
+        val userId = sharedPref.getString(key, "")
+        return userId!!
+    }
```

Comment:
The function signature of getUserEmail and getUserId is identical. I'd argue it's not the responsibility of the caller to know how preferences are working internally (and that it works with keys and it has to know which key results in which piece of data). Better keep the keys private to this class and just pass the right key for the right function